test/hotspot/jtreg/compiler/c2/irTests/TestScopedValue.java

```diff
@@ -0,0 +1,673 @@
+            return;
+        }
+        Method m = TestScopedValue.class.getDeclaredMethod(name, parameterTypes);
+        WHITE_BOX.enqueueMethodForCompilation(m, CompilerWhiteBoxTest.COMP_LEVEL_FULL_OPTIMIZATION);
+        if (!WHITE_BOX.isMethodCompiled(m) || WHITE_BOX.getMethodCompilationLevel(m) != CompilerWhiteBoxTest.COMP_LEVEL_FULL_OPTIMIZATION) {
+            throw new RuntimeException("should be compiled");
+        }
+    }
+
+    @DontInline
+    static void notInlined() {
+    }
+
+    @Test
+    @IR(failOn = {IRNode.CALL_OF_METHOD, "slowGet"})
+    @IR(counts = {IRNode.LOAD_D, "1" })
+    public static double testFastPath2() {
+        ScopedValue<MyDouble> scopedValue = sv;
+        MyDouble sv1 = scopedValue.get();
+        notInlined();
+        MyDouble sv2 = scopedValue.get(); // Should optimize out
+        return sv1.getValue() + sv2.getValue();
+    }
+
+    @Run(test = "testFastPath2", mode = RunMode.STANDALONE)
+    private void testFastPath2Runner() throws Exception {
+        ScopedValue.where(sv, new MyDouble(42)).run(
+                () -> {
+                    MyDouble unused = sv.get();
+                    for (int i = 0; i < 20_000; i++) {
+                        if (testFastPath2() != 42 + 42) {
+                            throw new RuntimeException();
+                        }
+                    }
+                });
+        forceCompilation("testFastPath2");
+    }
+
+    @Test
+    @IR(failOn = {IRNode.CALL_OF_METHOD, "slowGet"})
+    @IR(counts = {IRNode.LOAD_D, "2" })
+    public static double testFastPath3() {
+        MyDouble sv1 = sv.get();
+        notInlined();
+        MyDouble sv2 = sv.get(); // Doesn't optimize out (load of sv cannot common)
+        return sv1.getValue() + sv2.getValue();
+    }
+
+    @Run(test = "testFastPath3", mode = RunMode.STANDALONE)
+    private void testFastPath3Runner() throws Exception {
+        ScopedValue.where(sv, new MyDouble(42)).run(
+                () -> {
+                    MyDouble unused = sv.get();
+                    for (int i = 0; i < 20_000; i++) {
+                        if (testFastPath3() != 42 + 42) {
+                            throw new RuntimeException();
+                        }
+                    }
+                });
+        forceCompilation("testFastPath3");
+    }
+
+    // Split if test but it doesn't trigger at this point
+    // @Test
+    // @IR(failOn = {IRNode.CALL_OF_METHOD, "slowGet"})
+    // @IR(counts = {IRNode.LOAD_L, "2" })
+    // public static long test4(boolean flag) throws Exception {
+    //     ScopedValue<MyLong> scopedValue;
+    //     MyLong long1;
+    //     if (flag) {
+    //         scopedValue = svFinal;
+    //         long1 = (MyLong)svFinal.get();
+    //     } else {
+    //         scopedValue = svFinal2;
+    //         long1 = (MyLong)svFinal2.get();
+    //     }
+    //     MyLong long2 = (MyLong)scopedValue.get();
+    //     return long1.getValue() + long2.getValue();
+    // }
```

Comment:
Are you still working on this?

Reply:
No. I couldn't make the test work unfortunately, so I wasn't sure whether to leave the test commented out (in case someone revisits that later) or not.

Reply:
Maybe have the body of the test put in, and the IR-rules commented out, with a follow-up RFE for investigation, if you think there is something one can do about it?